### PR TITLE
Fix PS-2455 (nnodb does not report filename in I/O error message)

### PR DESCRIPTION
**Pull request by laurynas-biveinis**
Apply a patch by Marcelo Altmann at
https://github.com/percona/percona-server/pull/2919 which attempts to
get the file path from file descriptor through /proc/<pid>/fd/<fd>
symlink, present on Linux.

The original patch was changed to extract the /proc-reading code to a
separate function, which is used to report not only read but also
fsync and fallocate errors.

https://ps.cd.percona.com/view/5.7/job/percona-server-5.7-param/83/

### CODE REVIEW

File: storage/innobase/os/os0file.cc [hunk 1]
```diff
@@ -2462,6 +2462,18 @@ SyncFileIO::execute(const IORequest& request)
 	return(n_bytes);
 }
 
+MY_ATTRIBUTE((unused))
```
Comment (percona-ysorokin) [inline storage/innobase/os/os0file.cc]: `warn_unused_result` instead of `unused`?

File: storage/innobase/os/os0file.cc [hunk 2]
```diff
@@ -2489,11 +2501,20 @@ os_file_punch_hole_posix(
 		return(DB_IO_NO_PUNCH_HOLE);
 	}
 
-	ib::warn()
-		<< "fallocate(" << fh
-		<<", FALLOC_FL_PUNCH_HOLE | FALLOC_FL_KEEP_SIZE, "
-		<< off << ", " << len << ") returned errno: "
-		<<  errno;
+	const std::string fd_path = os_file_find_path_for_fd(fh);
+	if (fd_path != "") {
```
Comment (percona-ysorokin) [inline storage/innobase/os/os0file.cc]: `if (!fd_path.empty())`

**Review by percona-ysorokin — APPROVED** [review]
LGTM with `warn_unused_result` fixed

File: storage/innobase/os/os0file.cc [hunk 3]
```diff
@@ -3490,11 +3511,17 @@ os_file_fsync_posix(
...
 			break;
 
-		case EIO:
+		case EIO: {
 
-                        ib::fatal()
-				<< "fsync() returned EIO, aborting.";
+			const std::string fd_path
+				= os_file_find_path_for_fd(file);
+			if (fd_path != "")
```
Comment (percona-ysorokin) [inline storage/innobase/os/os0file.cc]: likewise

File: storage/innobase/os/os0file.cc [hunk 4]
```diff
@@ -6203,9 +6230,18 @@ os_file_read_page(
 			}
 		}
 
-		ib::error() << "Tried to read " << n
-			<< " bytes at offset " << offset
-			<< ", but was only able to read " << n_bytes;
+		const std::string fd_path = os_file_find_path_for_fd(file);
+		if (fd_path != "") {
```
Comment (percona-ysorokin) [inline storage/innobase/os/os0file.cc]: likewise

**Comment by laurynas-biveinis** [comment]
All comments updated, https://ps.cd.percona.com/view/5.7/job/percona-server-5.7-param/86/

**Pull request merged**